pinot-tools/src/main/java/com/linkedin/pinot/tools/admin/command/MoveReplicaGroup.java

```diff
@@ -0,0 +1,437 @@
+      String destinationServer = getDestinationServer(destinationServers, existingMapping);
+      if (destinationServer == null) {
+        throw new RuntimeException("No destination server for segment: " + segment);
+      }
+      String toRemove = null;
+
+      for (Map.Entry<String, String> existingInstanceEntry : existingMapping.entrySet()) {
+        if (srcHosts.contains(existingInstanceEntry.getKey())) {
+          toRemove = existingInstanceEntry.getKey();
+          break;
+        }
+      }
+
+      if (toRemove == null) {
+        throw new RuntimeException("Could not find a source host to remove for segment: " + segment);
+      }
+      existingMapping.remove(toRemove);
+      existingMapping.put(destinationServer, "ONLINE");
+    }
+    return newIdealState;
+  }
+
+  private String getDestinationServer(PriorityQueue<ServerInstance> destinationServers, Map<String, String> existingSegmentMapping) {
+    Preconditions.checkNotNull(destinationServers);
+    Preconditions.checkArgument(! destinationServers.isEmpty());
+
+    List<ServerInstance> removedServers = new ArrayList<>();
+    String selectedServer = null;
+    while (!destinationServers.isEmpty()) {
+      ServerInstance si = destinationServers.poll();
+      removedServers.add(si);
+      if (! existingSegmentMapping.containsKey(si.server)) {
+        selectedServer = si.server;
+        ++si.segments;
+        break;
+      }
+    }
+    for (ServerInstance removedServer : removedServers) {
+      destinationServers.add(removedServer);
+    }
+    return selectedServer;
+  }
+
+  private Map<String, Map<String, String>> copyIdealState(Map<String, Map<String, String>> idealStateMap) {
+    Map<String, Map<String, String>> copy = new HashMap<>(idealStateMap);
+    for (Map.Entry<String, Map<String, String>> segmentEntry : idealStateMap.entrySet()) {
+      Map<String, String> instanceCopy = new HashMap<>(segmentEntry.getValue().size());
+      for (Map.Entry<String, String> instanceEntry : segmentEntry.getValue().entrySet()) {
+        instanceCopy.put(instanceEntry.getKey(), instanceEntry.getValue());
+      }
+      copy.put(segmentEntry.getKey(), instanceCopy);
+    }
+    return copy;
+  }
+
+  private void verifyNoServerOverlap(List<String> srcHosts, List<String> taggedServers) {
+    for (String srcHost : srcHosts) {
+      if (taggedServers.contains(srcHost)) {
+        LOGGER.error("Source host: {} is also present in destination list", srcHost);
+        LOGGER.error("Refusing to migrate replica group");
+        System.exit(1);
+      }
+    }
+  }
+
+  private PriorityQueue<ServerInstance> getDestinationServerQueue(Map<String, Map<String, String>> idealStateMap,
+      List<String> destServers) {
+    // better to keep map rather than removing elements from heap each time
+    Map<String, ServerInstance> serverMap = new HashMap<>(destServers.size());
+    for (String ds : destServers) {
+      serverMap.put(ds, new ServerInstance(ds, 0));
```

Comment:
Oh, I got confused with the existing ServerInstance class. Looks like you have a new one in this file.